spec_util/meld.go

```diff
@@ -478,18 +480,42 @@ func structShouldBeMap(struc *pb.Struct) bool {
 	// A struct should be a map if its number of optional fields exceeds
 	// maxOptionalFieldsPerStruct.
 	numOptionalFields := 0
-	for _, field := range struc.Fields {
+	allFieldsStartWithNumbers := true
+	for fieldName, field := range struc.Fields {
 		if field.GetOptional() != nil {
 			numOptionalFields++
 			if numOptionalFields > maxOptionalFieldsPerStruct {
 				return true
 			}
 		}
+		if !startsWithNumber(fieldName) {
+			allFieldsStartWithNumbers = false
+		}
+	}
+
+	// A struct should be a map if all its fields start with numbers and there
+	// are a sufficient number of fields.  Because many programming languages
+	// disallow struct names starting with numbers, the number of fields is
+	// lower.
+	if allFieldsStartWithNumbers && len(struc.Fields) >= minNumberedFields {
+		return true
 	}
 
 	return false
 }
 
+func startsWithNumber(s string) bool {
+	if len(s) == 0 {
+		return false
+	}
+
+	if _, err := strconv.Atoi(s[0:1]); err != nil {
+		return false
+	}
+
+	return true
+}
```

Comment:
Why not use a regular expression `^\d`? Alternatively, could compare `'0' <= s[0] && s[0] <= '9'` instead of calling `Atoi`.